### PR TITLE
[Tune] Comet Integration

### PR DESCRIPTION
**Pull request by DN6**
## Why are these changes needed?
This PR adds a `CometLoggerCallback` to the Tune Integrations, allowing users to log runs from Ray to [Comet](https://www.comet.ml/site/). 

## Related issue number

This PR is based on the initial work started by @amogkam 
https://github.com/ray-project/ray/pull/15817

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by DN6** [comment]
@amogkam I've started the work on the PR. My colleague @mcullan will take over and create the tests and examples needed to get this merged in.

**Comment by amogkam** [comment]
Sounds good, thanks @DN6!

**Comment by amogkam** [comment]
FYI @krfricke @xwjiang2010

**Comment by mcullan** [comment]
@amogkam @krfricke @xwjiang2010 Hi all, I've been trying to figure out what would have caused these tests to fail this time. I don't think that they have any contact with the Comet integration and I didn't modify any other code, and moreover it looks like they're just failing due to timeouts. 

Do you have any thoughts on what I can try to do to sort this out? As well as next steps for getting this integration wrapped up? I think I need to add some documentation but I'm not sure how all that works. Do I need to add a file somewhere to tell the docs to include the docstrings I wrote in `comet.py`, for instance?

**Comment by xwjiang2010** [comment]
Hi Michael, this specific test is being flaky. I will take a look at that test separately. If this is the only one failing, you can mark the PR with @tests-ok label. Thanks!!

**Comment by mcullan** [comment]
Well it's the`mac-apple-medium-k-z` and `octopus-brain-tune-tests-and-examples-using-rllib` tests.  In `medium-k-z`, there are a few failures. In `octopus-brain`, I think it's just the `tune-test-api` test that's failing. 

But I think they're totally separate from my contributions, and possibly just the result of me rebasing onto master, so I'm going to add that label. Thanks so much!

**Comment by DN6** [comment]
Hi @xwjiang2010. Just following up here for visibility. I don't think Michael or I have the appropriate permissions to add the @tests-ok label to this PR.

**Comment by DN6** [comment]
Hi @xwjiang2010. Following up on this. I don't think @mcullan  or I have the appropriate permissions to add @tests-ok label to this PR. I think someone from the Ray team might have to do that.

File: python/ray/tune/integration/comet.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+            experiment = experiment_cls(**self.experiment_kwargs)
+            self._trial_experiments[trial] = experiment
+            # Set global experiment to None to allow for multiple experiments.
+            set_global_experiment(None)
+        else:
+            experiment = self._trial_experiments[trial]
+
+        experiment.set_name(str(trial))
+        experiment.add_tags(self.tags)
+
```
Comment (DN6) [inline python/ray/tune/integration/comet.py:142]:
```suggestion
experiment.log_other("Created from", "Ray")

```

**Comment by Yard1** [comment]
Hey @DN6 can you merge master into your branch? This will also rerun the tests. There also seem to be some lint issues.

**Comment by Yard1** [comment]
Hey @DN6 thanks for merging, the tests look fine. 

I have a couple of things to ask:
* Can you also add a documentation page? It would be awesome if we could get a short tutorial similar to the one we have for MLFlow - see https://ray--20766.org.readthedocs.build/en/20766/tune/tutorials/tune-mlflow.html. You will have to edit doc files. For the tutorial, you'd add a file like [this](https://github.com/ray-project/ray/blob/master/doc/source/tune/_tutorials/tune-mlflow.rst) and then link to it in [here](https://github.com/ray-project/ray/blob/master/doc/source/tune/api_docs/logging.rst).
* It would be great if we could also get a python script with a self-contained example - something like this https://github.com/ray-project/ray/blob/master/python/ray/tune/examples/mlflow_example.py. We also test those in CI by adding them to the BUILD file.

**Comment by DN6** [comment]
@Yard1 Sure thing. I can add documentation to this PR.  

There is self contained example here
https://github.com/ray-project/ray/blob/9cec51db3eae896cd0030d9a0b76d82eb515e8b7/python/ray/tune/examples/tune_comet_example.py
 
Should I add it to the BUILD file? What is the best way to include it?

**Comment by Yard1** [comment]
Sorry, missed the example! This should work for the build file in the tune folder (please add it in alphabetical order with other `_example`s):
```
py_test(
 name = "tune_comet_example",
 size = "medium",
 srcs = ["examples/tune_comet_example.py"],
 deps = [":tune_lib"],
 tags = ["team:ml", "exclusive", "example"]
)
```

**Comment by DN6** [comment]
Hi @Yard1. I've added the Comet tutorials for the docs. 

Before I push those up and trigger the CI again, I had one question, if I prepare an example Colab Notebook for the integration, would it be possible to link it somewhere in the docs? If so, could you point me to the appropriate place? 

There is also an issue that came up when I was trying to add an example to the BUILD file. In order to run the example, you will need a Comet API Key to be passed in. In that sense, it cannot be self contained. Is it alright to omit the example from the BUILD file?

File: python/ray/tune/examples/tune_comet_example.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    )
+    parser.add_argument(
+        "--project-name",
+        type=str,
+        help="Project to log experiment to. If "
+        "not passed in, experiment will be "
+        "logger under 'Uncategorized "
+        "Experiments'",
+    )
+    parser.add_argument(
```
Comment (Yard1) [inline python/ray/tune/examples/tune_comet_example.py:52]: Remove this unused argument please

**Comment by Yard1** [comment]
@DN6 I see, in this case it's probably fine to leave it out of the BUILD file, though it would be good to add a comment inside it explaining why this example is missing.

As for the colab notebook, it cannot be included in the docs directly, but you can include it as a specially formatted python script in the `doc/source/tune/_tutorials` folder. Eg. https://github.com/ray-project/ray/blob/master/doc/source/tune/_tutorials/tune-sklearn.rst will turn into https://docs.ray.io/en/latest/tune/tutorials/tune-sklearn.html. You will still have to link it in https://github.com/ray-project/ray/blob/master/doc/source/tune/_tutorials/overview.rst and https://github.com/ray-project/ray/blob/master/doc/source/tune/api_docs/logging.rst

**Comment by amogkam** [comment]
@DN6 for the example, you could follow what's being done for the `wandb_example` and mock the Comet API. Running a test with the mock API would still be useful to test our `CometLoggerCallback` API.

For the Colab notebook you could add it in this section https://docs.ray.io/en/latest/tune/tutorials/overview.html#colab-exercises

**Comment by Yard1** [comment]
Hey @DN6 just wanted to check up on this. Let me know if you need any support!

**Comment by DN6** [comment]
@Yard1 Sorry for the delay, wasn't able to get to this last week. I've added all the necessary changes for the documentation (links to examples, links in tutorials overview, link to  example Colab Notebook etc). I've also added the example in the BUILD file.  

I've added a flag so that the Comet example can run using a Mock API as suggested by @amogkam. I used the `wandb_example` as a reference. So it should be self contained. 

LMK if there's anything else you need from my side.

**Review by Yard1 — APPROVED** [review]
Looks good, just waiting on tests to pass!

**Comment by DN6** [comment]
Thanks for the help @Yard1 and for merging @amogkam! We're going to announce this integration on our community publications. Is there a schedule for the next release?

**Comment by Yard1** [comment]
We'll have a new release with this feature in about 2 weeks!

**Comment by DN6** [comment]
Hi @Yard1. 

I seems like the Comet Logger wasn't included in the latest [release](https://github.com/ray-project/ray/releases/tag/ray-1.10.0)? 

Following up to confirm if there's anything else that needs to be done on our side?

**Comment by Yard1** [comment]
Hey @DN6, unfortunately it didn't make the branch cut for this release. It will be included in the next one, ETA one week.

**Comment by DN6** [comment]
Hi @Yard1. Hope you're doing well. Just checking back in on this. Do you happen to know when the next release is scheduled?

**Comment by Yard1** [comment]
Should be today or tomorrow!
